src/Controls/src/Core/HandlerImpl/NavigationPage/NavigationPage.Impl.cs

```diff
@@ -110,62 +128,102 @@ void OnAppearing(object sender, EventArgs e)
 				window.Toolbar.ApplyNavigationPage(this);
 		}
 
-		Task WaitForCurrentNavigationTask() =>
-			CurrentNavigationTask ?? Task.CompletedTask;
-
 		// This is used for navigation events that don't effect the currently visible page
 		// InsertPageBefore/RemovePage
 		async void SendHandlerUpdate(bool animated)
 		{
-			await WaitForCurrentNavigationTask();
-			var trulyReadOnlyNavigationStack = new List<IView>(NavigationStack);
-			var request = new NavigationRequest(trulyReadOnlyNavigationStack, animated);
-			((INavigationView)this).RequestNavigation(request);
+			try
+			{
+				Interlocked.Increment(ref _waitingCount);
+				await SemaphoreSlim.WaitAsync();
+				var trulyReadOnlyNavigationStack = new List<IView>(NavigationStack);
+				var request = new NavigationRequest(trulyReadOnlyNavigationStack, animated);
+				((INavigationView)this).RequestNavigation(request);
+			}
+			finally
+			{
+				Interlocked.Decrement(ref _waitingCount);
+				SemaphoreSlim.Release();
+			}
 		}
 
-		TaskCompletionSource<object> _taskCompletionSource;
-
-		async Task SendHandlerUpdateAsync(bool animated, bool push = false, bool pop = false, bool popToRoot = false)
+		async Task SendHandlerUpdateAsync(
+			bool animated,
+			Action processStackChanges,
+			Action firePostNavigatingEvents,
+			Action fireNavigatedEvents)
 		{
-			// Wait for any pending navigation tasks to finish
-			await WaitForCurrentNavigationTask();
+			try
+			{
+				processStackChanges?.Invoke();
+				Interlocked.Increment(ref _waitingCount);
+				// Wait for pending navigation tasks to finish
+				await SemaphoreSlim.WaitAsync();
 
-			var completionSource = new TaskCompletionSource<object>();
-			CurrentNavigationTask = completionSource.Task;
-			_taskCompletionSource = completionSource;
+				var currentNavRequestTaskSource = new TaskCompletionSource<object>();
```

Comment:
I took this suggestion from [here](https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#always-create-taskcompletionsourcet-with-taskcreationoptionsruncontinuationsasynchronously) but since this is a UI framework, not sure if this is good or bad addition.
```suggestion
				var currentNavRequestTaskSource = new TaskCompletionSource<object>(TaskCreationOptions.RunContinuationsAsynchronously);
```

Reply:
@pictos can you create an issue with this suggestion? There are a few places where we use TaskCompletionSource and it's going to need a bit of a deeper investigation before we switch over to those options. We'll need to figure out if those suggestions make sense for mobile workloads.

Reply:
@PureWeen done 
- #3347